### PR TITLE
Use CMake extensions repo

### PR DESCRIPTION
**Pull request by CedricGuillemet**
This PR needs https://github.com/BabylonJS/UrlLib/pull/6 for Unix websocket stub impl.

### CODE REVIEW

File: Core/Node-API/CMakeLists.txt [hunk 1]
```diff
@@ -15,12 +15,7 @@ if(NAPI_BUILD_ABI)
             message(STATUS "Installing ${name}-android module")
             file(COPY "${CMAKE_CURRENT_SOURCE_DIR}/package-${name}.json" DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
             file(RENAME "${CMAKE_CURRENT_BINARY_DIR}/package-${name}.json" "${CMAKE_CURRENT_BINARY_DIR}/package.json")
-            execute_process(COMMAND ${NPM_COMMAND} install --silent --no-package-lock WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR} RESULT_VARIABLE NPM_RESULT)
-            if(NPM_RESULT EQUAL 0)
-                message(STATUS "Installing ${name}-android module - done")
```
Comment (bghgary) [inline Core/Node-API/CMakeLists.txt:20]: How do we keep this message?

File: Core/Node-API/CMakeLists.txt [hunk 2]
```diff
@@ -74,6 +70,9 @@ if(NAPI_BUILD_ABI)
             find_library(JAVASCRIPTCORE_LIBRARY JavaScriptCore)
             set(LINK_LIBRARIES ${LINK_LIBRARIES}
                 PUBLIC ${JAVASCRIPTCORE_LIBRARY})
+        elseif(UNIX)
+            set(LINK_LIBRARIES ${LINK_LIBRARIES}
+                PUBLIC ${JSCORE_LIBRARY})
```
Comment (bghgary) [inline Core/Node-API/CMakeLists.txt:75]: nit: to be consistent with APPLE
```suggestion
                PUBLIC ${JAVASCRIPTCORE_LIBRARY})
```

File: Core/Node-API/CMakeLists.txt [hunk 3]
```diff
@@ -114,3 +113,7 @@ target_link_libraries(napi ${LINK_LIBRARIES})
 
 set_property(TARGET napi PROPERTY FOLDER Dependencies)
 source_group(TREE ${CMAKE_CURRENT_SOURCE_DIR} FILES ${SOURCES})
+
+if(UNIX AND NOT APPLE AND NOT ANDROID)
+    target_include_directories(napi PUBLIC "/usr/include/webkitgtk-4.0/")
+endif()
```
Comment (bghgary) [inline Core/Node-API/CMakeLists.txt:119]: Use INCLUDE_DIRECTORIES and add it in the UNIX section above.

File: Tests/UnitTests/Android/app/build.gradle [hunk 4]
```diff
@@ -70,25 +70,37 @@ dependencies {
...
+            from "../../../node_modules/mocha"
+            include "mocha.js"
+            into 'src/main/assets/Scripts'
+        }
+        copy
+        {
+            from '../../Scripts'
+            include "*.js"
+            into 'src/main/assets/Scripts'
+        }
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/build.gradle:93]: Are we running `npm install` in the source folder? Ideally, we should use the CMake binary folder.

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt [hunk 5]
```diff
@@ -8,14 +8,17 @@ project(UnitTestsJNI)
 get_filename_component(UNIT_TESTS_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../../.." ABSOLUTE)
 get_filename_component(REPO_ROOT_DIR "${UNIT_TESTS_DIR}/../.." ABSOLUTE)
 
+include("${UNIT_TESTS_DIR}/../../content.cmake")
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt:11]:
```suggestion
include("${REPO_ROOT_DIR}/content.cmake")
```

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt [hunk 6]
```diff
@@ -8,14 +8,17 @@ project(UnitTestsJNI)
...
+npm(install "${UNIT_TESTS_DIR}/../" "UnitTests")
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt:12]: I don't like this. It would be better if we didn't pollute the source dir and instead do it in the CMake binary dir.

File: Core/Node-API-JSI/CMakeLists.txt [hunk 7]
```diff
@@ -68,3 +68,7 @@ endif()
 
 set_property(TARGET napi PROPERTY FOLDER Dependencies)
 source_group(TREE ${CMAKE_CURRENT_SOURCE_DIR} FILES ${SOURCES})
+
+if(UNIX AND NOT APPLE AND NOT ANDROID)
+    target_include_directories(napi PUBLIC "/usr/include/webkitgtk-4.0/")
+endif()
```
Comment (bghgary) [inline Core/Node-API-JSI/CMakeLists.txt:74]: Why is this here?

File: Core/Node-API/CMakeLists.txt [hunk 8]
```diff
@@ -3,8 +3,10 @@ if(WIN32)
     set(NAPI_JAVASCRIPT_ENGINE "Chakra" CACHE STRING "JavaScript engine for Node-API")
 elseif(APPLE)
     set(NAPI_JAVASCRIPT_ENGINE "JavaScriptCore" CACHE STRING "JavaScript engine for Node-API")
-elseif(UNIX)
+elseif(ANDROID)
     set(NAPI_JAVASCRIPT_ENGINE "V8" CACHE STRING "JavaScript engine for Node-API")
+elseif(UNIX)
+    set(NAPI_JAVASCRIPT_ENGINE "JavaScriptCore" CACHE STRING "JavaScript engine for Node-API")
```
Comment (bghgary) [inline Core/Node-API/CMakeLists.txt:9]: Linux doesn't support V8? I wonder if we should be defaulting to V8?

File: Tests/UnitTests/Android/app/build.gradle [hunk 9]
```diff
@@ -70,25 +70,37 @@ dependencies {
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.5.1'
 }
 
-task copyScripts {
-    copy
-    {
-        from "../../../node_modules/chai"
-        include "chai.js"
-        into 'src/main/assets/Scripts'
+task copyScripts{
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/build.gradle:73]:
```suggestion
task copyScripts {
```

File: Tests/UnitTests/CMakeLists.txt [hunk 10]
```diff
@@ -78,5 +81,6 @@ else()
 endif()
 
 set_property(TARGET UnitTests PROPERTY FOLDER Tests)
+add_definitions(-DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
```
Comment (bghgary) [inline Tests/UnitTests/CMakeLists.txt:84]: ditto

File: Tests/UnitTests/Shared/Shared.cpp [hunk 11]
```diff
@@ -43,6 +43,8 @@ int RunTests(Babylon::Polyfills::Console::CallbackT consoleCallback)
             Napi::Env env = info.Env();
             exitCode.set_value(info[0].As<Napi::Number>().Int32Value());
         }, "SetExitCode"));
+
+        env.Global().Set("RuntimeHostPlatform", Napi::Value::From(env, JSRUNTIMEHOST_PLATFORM));
```
Comment (bghgary) [inline Tests/UnitTests/Shared/Shared.cpp:47]:
```suggestion
        env.Global().Set("hostPlatform", Napi::Value::From(env, JSRUNTIMEHOST_PLATFORM));
```

I also think maybe we should add both `setExitCode` and this to a new object `test` maybe? And use camelCase to be consistent with other JS objects.

File: content.cmake [hunk 12]
```diff
@@ -0,0 +1,24 @@
+# fetch necessary modules
+
+include(FetchContent)
+
+# CMakeExtensions
+include(FetchContent)
+FetchContent_Declare(cmake-extensions
+        GIT_REPOSITORY https://github.com/BabylonJS/CMakeExtensions.git
+        GIT_TAG 7bdd0664181ce3d90dc94474464cea9b25b22db7)
```
Comment (bghgary) [inline content.cmake:9]: nit
```suggestion
    GIT_REPOSITORY https://github.com/BabylonJS/CMakeExtensions.git
    GIT_TAG 7bdd0664181ce3d90dc94474464cea9b25b22db7)
```

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt (hunk 5 above)
Comment (bghgary) [inline Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt:11]: `content.cmake` includes iOS stuff which needs to happen before project. It's a bit weird to factor it out this way.

File: CMakeLists.txt [hunk 13]
```diff
@@ -5,22 +5,7 @@ if (CMAKE_VERSION VERSION_GREATER_EQUAL "3.24.0")
...
-    message(STATUS "Fetching ios-cmake")
-    FetchContent_MakeAvailable(ios-cmake)
-    message(STATUS "Fetching ios-cmake - done")
-
-    set(CMAKE_TOOLCHAIN_FILE "${ios-cmake_SOURCE_DIR}/ios.toolchain.cmake" CACHE PATH "")
-    set(PLATFORM "OS64COMBINED" CACHE STRING "")
-    set(DEPLOYMENT_TARGET "12" CACHE STRING "")
-endif()
-
+include(content.cmake)
```
Comment (bghgary) [inline CMakeLists.txt:8]: See other comment.

File: content.cmake [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+FetchContent_MakeAvailable(cmake-extensions)
```
Comment (bghgary) [inline content.cmake:10]:
```suggestion
message(STATUS "Fetching CMakeExtensions")
FetchContent_MakeAvailable(cmake-extensions)
message(STATUS "Fetching CMakeExtensions - done")
```

File: Core/Node-API/CMakeLists.txt (hunk 2 above)
Comment (CedricGuillemet) [reply Core/Node-API/CMakeLists.txt]: Fixed.

File: Tests/UnitTests/Android/app/build.gradle [hunk 15]
```diff
@@ -26,7 +26,7 @@ android {
                 arguments (
                     "-DANDROID_STL=c++_shared",
                     "-DNAPI_JAVASCRIPT_ENGINE=${jsEngine}",
-                    "-DJSRUNTIMEHOST_CORE_APPRUNTIME_V8_INSPECTOR=ON"
+                    "-DJSRUNTIMEHOST_CORE_APPRUNTIME_V8_INSPECTOR=OFF"
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/build.gradle]: Fixed.

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt (hunk 5 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt]: Fixed.

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt [hunk 16]
```diff
@@ -8,14 +8,17 @@ project(UnitTestsJNI)
...
+
 set(JSRUNTIMEHOST_TESTS OFF) # Turn off the tests folder for Android Studio path
 add_subdirectory(${REPO_ROOT_DIR} "${CMAKE_CURRENT_BINARY_DIR}/JsRuntimeHost")
 
 add_library(UnitTestsJNI SHARED
     JNI.cpp
     ${UNIT_TESTS_DIR}/Shared/Shared.h
     ${UNIT_TESTS_DIR}/Shared/Shared.cpp)
-
+add_definitions(-DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt]: Fixed.

File: .github/jobs/linux.yml [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+  pool:
+    vmImage: ubuntu-latest
+
+  steps:
+  - script: |
+      sudo apt-get update
+      sudo apt-get install libjavascriptcoregtk-4.0-dev libcurl4-openssl-dev ninja-build
+    displayName: 'Install packages'
+
+  - script: cmake -B Build/ubuntu -GNinja -DJSCORE_LIBRARY=/usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.0.so -DNAPI_JAVASCRIPT_ENGINE=JavaScriptCore -DCMAKE_BUILD_TYPE=RelWithDebInfo
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply .github/jobs/linux.yml]: Fixed.

File: Tests/UnitTests/Android/app/build.gradle (hunk 9 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/build.gradle]: Fixed.

File: Tests/UnitTests/CMakeLists.txt (hunk 10 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/CMakeLists.txt]: Fixed.

File: content.cmake (hunk 12 above)
Comment (CedricGuillemet) [reply content.cmake]: Fixed.

File: content.cmake (hunk 14 above)
Comment (CedricGuillemet) [reply content.cmake:11]: Fixed.

File: Core/Node-API/CMakeLists.txt (hunk 8 above)
Comment (CedricGuillemet) [reply Core/Node-API/CMakeLists.txt:9]: There is a recent V8 package that has not been tested. I have no opinion on Linux default. Linux build here is very fast (1minute). We can test both JSC and V8. Issue opened here : https://github.com/BabylonJS/JsRuntimeHost/issues/34

File: Tests/UnitTests/Shared/Shared.cpp (hunk 11 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Shared/Shared.cpp]: Adding a new object for a value and a function in this test context looks like over engineering to me. Maybe if it grows (a lot) more. When Unix websocket is available and if test doesn't change, then `hostPlatform` will be removed.

File: CMakeLists.txt (hunk 13 above)
Comment (CedricGuillemet) [reply CMakeLists.txt:8]: Let's talk offline

File: Tests/UnitTests/Android/app/build.gradle (hunk 4 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/build.gradle:93]: Let's talk offline.

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt [hunk 18]
```diff
@@ -18,7 +19,7 @@ add_library(UnitTestsJNI SHARED
     JNI.cpp
     ${UNIT_TESTS_DIR}/Shared/Shared.h
     ${UNIT_TESTS_DIR}/Shared/Shared.cpp)
-add_definitions(-DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
+target_compile_definitions(UnitTestsJNI PUBLIC -DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt:22]: Should this be public? Also, `-D` isn't necessary.

```suggestion
target_compile_definitions(UnitTestsJNI PRIVATE JSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
```

File: Tests/UnitTests/Android/app/build.gradle (hunk 4 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/build.gradle:93]: Not doing it in this PR. Discussion captured here : https://github.com/BabylonJS/JsRuntimeHost/issues/36

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt (hunk 18 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt]: Fixed.

File: Tests/UnitTests/CMakeLists.txt [hunk 19]
```diff
@@ -81,6 +81,7 @@ else()
 endif()
 
 set_property(TARGET UnitTests PROPERTY FOLDER Tests)
-add_definitions(-DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
+target_compile_definitions(UnitTests PUBLIC -DJSRUNTIMEHOST_PLATFORM="${JSRUNTIMEHOST_PLATFORM}")
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Tests/UnitTests/CMakeLists.txt]: Fixed.

File: Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt (hunk 5 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/src/main/cpp/CMakeLists.txt]: include moved.

File: CMakeLists.txt (hunk 13 above)
Comment (CedricGuillemet) [reply CMakeLists.txt:8]: include moved.

Comment (CedricGuillemet) [reply CMakeLists.txt]: Fixed by moving everyfetch content in root cmake.

**Review by bghgary — COMMENTED** [review]
LGTM

**Review by bghgary — APPROVED** [review]
